php/libraries/OutputWrapper.class.inc

```
@@ -0,0 +1,188 @@
+     * @param ?SPLFileInfo $logDirectory The directory that will contain log
+     *                                   files written by this script.
+     *
+     * @return void
+     */
+    public function __construct(SPLFileInfo $logDirectory = null)
+    {
+        $this->_logDirectory = new SPLFileInfo(
+            $logDirectory ?? __DIR__ . '/tools/'
+        );
+    }
+
+    /**
+     * Print a formatted and indented bullet point.
+     *
+     * @param string $line The output.
+     *
+     * @return void
+     */
+    public function printBulletPoint(string $line): void
+    {
+        echo "\t* $line\n";
+    }
+
+    /**
+     * Prints a line of text with a newline before. Can optionally be printed
+     * to the command-line in coloured text. Valid options for colours are given
+     * in the COLOURS class constant.
+     * If logging is enabled, the line will also be written to a log file. Log
+     * files will not contain coloured text.
+     *
+     * @param string   $line   To be written.
+     * @param string   $colour A valid colour, given by COLOURS, for the output.
+     * @param resource $handle A file resource where text should be written.
+     *
+     * @return void
+     */
+    public function printLine(
+        string $line,
+        string $colour = '',
+        $handle = STDOUT
+    ) {
+        $output = $line;
+
+        // Verify that if a colour is supplied, it is valid
+        if (!empty($colour)
+            && !array_key_exists(strtolower($colour), self::COLOURS)
+        ) {
+            throw new \InvalidArgumentException(
+                "Invalid argument for colour (`$colour`)"
+            );
+        }
+
+        // If a valid colour is supplied and the output is going to a TTY,
+        // colourize the output.
+        if (!empty($colour) && stream_isatty($handle)) {
+            // Format the output to colour the whole line.
+            $output = sprintf("\n\e[%sm$line\e[0m", self::COLOURS[$colour]);
+        }
```

this needs an else, otherwise it'll silently eat the line if not attached to a TTY.

---

`$output` gets initialized above. The if statement modifies `$output` if `$handle` is a TTY. In every case `$output` is written to `$handle` on line 96 so nothing will get silently eaten.

---

my mistake.. I was misreading that as printf not sprintf.